networkx/algorithms/d_separation.py

```diff
@@ -0,0 +1,121 @@
+"""
+Algorithm for testing d-separation in DAGs.
+
+*d-separation* is a test for conditional independence in probability
+distributions that can be factorized using DAGs.  It is a purely
+graphical test that uses the underlying graph and makes no reference
+to the actual distribution parameters.  For a detailed treatment see
+[1]_, [2]_.
+"""
+
+from collections import deque
+from typing import Hashable, AbstractSet
+
+import networkx as nx
+from networkx.utils import not_implemented_for, UnionFind
+
+__all__ = ["d_separated"]
+
+
+@not_implemented_for("undirected")
+def d_separated(G: nx.DiGraph, x: AbstractSet[Hashable],
```

Comment:
@ericmjl While these type hints may be useful, the function signature in the documentation ends up being cluttered. I am leaning towards removing the type hints since the types are mostly documented in the docstring below. I would like to hear your thoughts on that.

Reply:
@arunwise they're actually a very good practice, and I was super happy you put them in there in the first place! 😄 

This is because there are static code checkers like mypy that leverage type hints to check that the code written is correct. For example, with mypy on, I can (in source `.py` files) know whether I've passed in an object that is not of the categories accepted by a function. I'd be inclined to leave them in here, and if you really would like to remove some clutter, taking them out from the docstring could be a way out, though I'd leave those in there as well, because they are the "human-readable" form of the docs.